change guard_or impl for better perf and simplicity

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153674


PR time benchmarks has been showing regressions as we move to guard_or_false, reason is that prev implementation do not cache. 
This new approach will propagate the fallback value to eval and return it. allowing eval to cache and reducing scamming logs and complexity. 

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

laithsakka committed:
change guard_or impl for better perf and simplicity

[ghstack-poisoned]

bobrenjc93 commented on test/test_dynamic_shapes.py:
```diff
@@ -2848,7 +2848,7 @@ def func(a, b):
                 return b * 20
 
-        # call with guarding.
+        # eager.
```
i'm not sure how helpful this comment is, it seems self evident from the code?

bobrenjc93 reviewed (changes requested):
overall looks reasonable to me, back to you for test failures

laithsakka committed:
Update on "change guard_or impl for better perf and simplicity"



PR time benchmarks has been showing regressions as we move to guard_or_false, reason is that prev implementation do not cache. 
This new approach will propagate the fallback value to eval and return it. allowing eval to cache and reducing scamming logs and complexity. 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -6941,17 +6910,25 @@ def evaluate_expr(
     ) -> sympy.Basic:
         """
         Given an expression, evaluates it, adding guards if necessary
+        When fallback_value is not None the function return fallback_value instead of failing with data dependent errror
```
errror -> error

laithsakka commented:
failures are not related.

laithsakka commented:
@pytorchbot rebase

laithsakka committed:
Update

[ghstack-poisoned]

bobrenjc93 reviewed (approved):
please address my comment before landing

bobrenjc93 commented on test/test_dynamic_shapes.py:
```diff
@@ -2913,7 +2913,7 @@ def func(a, b):
                 return b * 20
 
-        # call with guarding.
+        # eager.
```
rename test to be test_eager_guard_or_false?

laithsakka replied:
i did this 
```

        # eager.
        self.assertEqual(func(torch.tensor([1]), torch.tensor([1])), torch.tensor([10]))
        self.assertEqual(func(torch.tensor([2]), torch.tensor([1])), torch.tensor([20]))

        # compile with unbacked.
        unbacked_func = torch.compile(func, dynamic=True, fullgraph=True)
        a = torch.tensor([1])
        b = torch.tensor([1])
        unbacked_func(a, b)
   ```

laithsakka committed:
Update on "change guard_or impl for better perf and simplicity"



PR time benchmarks has been showing regressions as we move to guard_or_false, reason is that prev implementation do not cache. 
This new approach will propagate the fallback value to eval and return it. allowing eval to cache and reducing scamming logs and complexity. 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka commented:
address comments and rebase on main

laithsakka committed:
Update on "change guard_or impl for better perf and simplicity"



PR time benchmarks has been showing regressions as we move to guard_or_false, reason is that prev implementation do not cache. 
This new approach will propagate the fallback value to eval and return it. allowing eval to cache and reducing scamming logs and complexity. 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka committed:
Update on "change guard_or impl for better perf and simplicity"



PR time benchmarks has been showing regressions as we move to guard_or_false, reason is that prev implementation do not cache. 
This new approach will propagate the fallback value to eval and return it. allowing eval to cache and reducing scamming logs and complexity. 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka committed:
Update on "change guard_or impl for better perf and simplicity"



PR time benchmarks has been showing regressions as we move to guard_or_false, reason is that prev implementation do not cache. 
This new approach will propagate the fallback value to eval and return it. allowing eval to cache and reducing scamming logs and complexity. 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge -f "just updated expected result after no faluires"
